test/rateLimiters/slidingWindowCounter.test.ts

```diff
@@ -0,0 +1,476 @@
+            expect(count.previousTokens).toBe(initRequest);
+        });
+    });
+
+    describe('SlidingWindowCounter functions as expected', () => {
+        afterEach(() => {
+            client.flushall();
+        });
+
+        test('allows user to consume current allotment of tokens', async () => {
+            // "free requests"
+            expect((await limiter.processRequest(user1, timestamp, 0)).success).toBe(true);
+            // Test 1 token requested
+            expect((await limiter.processRequest(user1, timestamp, 1)).success).toBe(true);
+            // Test < CAPACITY tokens requested
+            expect((await limiter.processRequest(user2, timestamp, CAPACITY - 1)).success).toBe(
+                true
+            );
+            // <= CAPACITY tokens requested
+            expect((await limiter.processRequest(user3, timestamp, CAPACITY)).success).toBe(true);
+        });
+
+        test('blocks exceeding requests over token allotment', async () => {
+            // Test > capacity tokens requested
+            expect((await limiter.processRequest(user1, timestamp, CAPACITY + 1)).success).toBe(
+                false
+            );
+
+            // Fill up user 1's window
+            const value: RedisWindow = {
+                currentTokens: 10,
+                previousTokens: null,
+                fixedWindowStart: timestamp,
+            };
+            await client.set(user1, JSON.stringify(value));
+
+            // window is full. Shouldn't be allowed to take 1 token
+            expect((await limiter.processRequest(user1, timestamp, 1)).success).toBe(false);
+
+            // Should still be allowed to process "free" requests
+            expect((await limiter.processRequest(user1, timestamp, 0)).success).toBe(true);
+        });
+
+        test('fixed window and current/previous tokens update as expected', async () => {
+            await setTokenCountInClient(client, user1, 0, null, timestamp);
+            // fills first window with 4 tokens
+            await limiter.processRequest(user1, timestamp, 5);
+            // fills second window with 5 tokens
+            expect(
+                await (
+                    await limiter.processRequest(user1, timestamp + WINDOW_SIZE + 1, 4)
+                ).tokens
+            ).toBe(6);
+            // currentTokens (in current fixed window): 0
+            // previousTokens (in previous fixed window): 8
+            const count = await getWindowFromClient(client, user4);
+            // ensures that fixed window is updated when a request goes over
+            expect(count.fixedWindowStart).toBe(timestamp + WINDOW_SIZE);
+            // ensures that previous tokens property updates on fixed window change
+            expect(count.previousTokens).toBe(5);
+            // ensures that current tokens only represents tokens from current window requests
+            expect(count.currentTokens).toBe(4);
+        });
+
```

Comment:
What is the expected behavior if a request comes in after more than `WINDOW_SIZE` milliseconds? Let's add a testcase for this (unless I missed it).